### PR TITLE
akka-stream compiling on Scala 3

### PR DESCRIPTION
**Pull request by johanandren**
Somewhat Refs #30243

Superseedes #30174 and #30223

Requires Scala 3.0.1-RC1 so not _entirely_ ready to be merged. Everything compiles and no problematic bin-comp issues as far as I can see though.

### CODE REVIEW

**Comment by johanandren** [comment]
BTW I created this as a completely new set of commits, based on the two previous tries, so I'm afraid @dwijnand will not get attribution for his very valuable work to move us forward with this. Not sure if we can do something about that, perhaps at least a shoutout in the release notes once this is in.

**Comment by dwijnand** [comment]
You picking up and making use of the work I put into this is enough for me to be grateful. ❤️

**Review by andreaTP — APPROVED** [review]

**Comment by johanandren** [comment]
Yeah, I guess as long as we don't release any artifacts, perhaps a RC dependency is only in the build and fine.

File: akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala [hunk 1]
```diff
@@ -178,7 +178,7 @@ import akka.util.ByteString
 
       private def unbindCompleted(): Unit = {
         stageActor.unwatch(listener)
-        unbindPromise.trySuccess(Done)
+        unbindPromise.trySuccess(())
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala:181]: fascinating that Scala 2 allows that when reading it like this, but not so strange when thinking some more about it

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**